pkg/ingester/activeseries/active_series.go

```diff
@@ -65,12 +65,12 @@ type seriesStripe struct {
 
 	mu                                   sync.RWMutex
 	refs                                 map[storage.SeriesRef]seriesEntry
-	active                               uint32   // Number of active entries in this stripe. Only decreased during purge or clear.
-	activeMatching                       []uint32 // Number of active entries in this stripe matching each matcher of the configured Matchers.
-	activeNativeHistograms               uint32   // Number of active entries (only native histograms) in this stripe. Only decreased during purge or clear.
-	activeMatchingNativeHistograms       []uint32 // Number of active entries (only native histograms) in this stripe matching each matcher of the configured Matchers.
-	activeNativeHistogramBuckets         uint32   // Number of buckets in active native histogram entries in this stripe. Only decreased during purge or clear.
-	activeMatchingNativeHistogramBuckets []uint32 // Number of buckets in active native histogram entries in this stripe matching each matcher of the configured Matchers.
+	active                               uint32            // Number of active entries in this stripe. Only decreased during purge or clear.
+	activeMatching                       map[uint16]uint32 // Number of active entries in this stripe matching each matcher of the configured Matchers.
```

Comment:
Actually in the past we _did_ use a map, and then we switched to a slice. The reason behind this is that usually a series only matches between zero and one tracker (the trackers are usually the integration where they come from) with a few more in some use cases, but never _many_.

Reply:
The motivation for this is https://github.com/grafana/mimir/pull/5318#issuecomment-1669238893 and https://raintank-corp.slack.com/archives/C04AT2ANL90/p1690989928160289

Before @bboreham 's recent change in https://github.com/grafana/mimir/pull/5665 it was taking up a lot of memory, and even afterwards we are still tripling the amount needed as we are tracking native histogram series and buckets as well. Discussed with @krajorama and we do need to track those, and it makes sense to use the same set of custom matchers for float series and native histogram series, so we are exploring other ways to minimise the memory usage especially in cases where native histograms are not being used. So we could just implement the change for native histograms and leave the float active series tracking alone, but we thought we might as well explore the possibility of making them consistent.

Benchmark results:
[compare_BenchmarkActiveSeries_UpdateSeries$.txt](https://github.com/grafana/mimir/files/12468121/compare_BenchmarkActiveSeries_UpdateSeries.txt)
[compare_BenchmarkMatchesSeries.txt](https://github.com/grafana/mimir/files/12468124/compare_BenchmarkMatchesSeries.txt)

Reply:
Note in your benchmarks that `UpdateSeries` is 5% slower in meaningful cases (when there are series, with or without matchers) while there's no improvement for the meaningful cases (there's only improvement when there are no series, saving 150Ki, same order of magnitude that I made above, but when there are 100K series the saving is 0.01% (because it's constant, doesn't depend on the series amount). There's also an increase in allocations, which will slow down the process more (as it generates more garbage to collect).

I wouldn't trade a 5% speed (which stretches with number of series) for 150Ki of constant memory.